### ISSUE TITLE
npm install loom-js gives error

### ISSUE THREAD

**Issue by shashi278** [loomnetwork/loom-js#345]
Running `npm install loom-js` gives below error:
```bash
npm ERR! code 128
npm ERR! git dep preparation failed
npm ERR! command /usr/bin/node /usr/local/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/home/watney/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit
npm ERR! npm WARN using --force Recommended protections disabled.
npm ERR! npm ERR! code 128
npm ERR! npm ERR! command failed
npm ERR! npm ERR! command git clone --mirror -q ssh://git@github.com/gulpjs/gulp.git /home/watney/.npm/_cacache/tmp/git-clone-0e411a93/.git
npm ERR! npm ERR! fatal: destination path '/home/watney/.npm/_cacache/tmp/git-clone-0e411a93/.git' already exists and is not an empty directory.
npm ERR! 
npm ERR! npm ERR! A complete log of this run can be found in:
npm ERR! npm ERR!     /home/watney/.npm/_logs/2021-04-21T14_00_51_817Z-debug.log

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/watney/.npm/_logs/2021-04-21T14_00_52_530Z-debug.log
```
`node -v`: v12.22.1
`npm -v`: 7.10.0

I even tried cleaning cache `npm cache clean --force` but it did not help :disappointed: 

Is it a known issue? How can I solve this?

**Comment by enlight** [comment]
This seems to be a bug in NPM rather that anything to do with `loom-js`, only thing I can suggest is using `yarn install` instead, that's what we use internally.

**Comment by shashi278** [comment]
Thanks for responding but that didn't help either. In fact, I was getting an even longer error with yarn.
I had to upgrade node to `v14.16.1`. But again I was getting some issue related to `scrypt@6.0.3` then I had to npm install `github:barrysteyn/node-scrypt#fb60a8d3c158fe115a624b5ffa7480f3a24b03fb` which finally solved the above issue.

**Comment by joshuaebowling** [comment]
same issue. initially posted this to loom-truffle-provider only to realize that loom-js is the issue. Here's the contents of that ticket.
Like OP, I've tried both yarn and npm.



Node version: 14.16.1
node-gyp: 5.1.0

I tried to install: npm install loom-truffle-provider and received the following error:
Error: Error: Command failed: ./configure
configure: error: in /home/josh/.nvm/versions/node/v14.16.1/lib/node_modules/loom-truffle-provider/node_modules/scrypt/scrypt/scrypt-1.2.0': configure: error: no acceptable C compiler found in $PATH See config.log' for more details

so, to me, it appeared I needed to install build-tools for linux, which I did.

then, I tried to install the package again. Now, I get a different error, apparently indicating that some calls to some V8 methods that appear to have different call signatures. From this, I suspect that maybe a need a different version of node. Can anyone confirm? Or, disconfirm and help me understand the best way to get this package installed?

thanks!

Error log folows:
In file included from ../src/node-boilerplate/scrypt_common.cc:28:
../../nan/nan.h: In function ‘void Nan::AsyncQueueWorker(Nan::AsyncWorker*)’:
../../nan/nan.h:2232:62: warning: cast between incompatible function types from ‘void ()(uv_work_t)’ {aka ‘void ()(uv_work_s)’} to ‘uv_after_work_cb’ {aka ‘void ()(uv_work_s, int)’} [-Wcast-function-type]
2232 | , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
| ^
In file included from ../src/node-boilerplate/scrypt_params_async.cc:1:
../../nan/nan.h: In function ‘void Nan::AsyncQueueWorker(Nan::AsyncWorker*)’:
../../nan/nan.h:2232:62: warning: cast between incompatible function types from ‘void ()(uv_work_t)’ {aka ‘void ()(uv_work_s)’} to ‘uv_after_work_cb’ {aka ‘void ()(uv_work_s, int)’} [-Wcast-function-type]
2232 | , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
| ^
In file included from ../src/node-boilerplate/inc/scrypt_async.h:28,
from ../src/node-boilerplate/inc/scrypt_params_async.h:28,
from ../src/node-boilerplate/scrypt_params_async.cc:4:
../src/node-boilerplate/inc/scrypt_common.h: In constructor ‘NodeScrypt::Params::Params(const v8::Localv8::Object&)’:
../src/node-boilerplate/inc/scrypt_common.h:39:48: error: no matching function for call to ‘v8::Object::Get(v8::Localv8::String)’
39 | N(obj->Get(Nan::New("N").ToLocalChecked())->Uint32Value()),
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3717:43: note: candidate: ‘v8::MaybeLocalv8::Value v8::Object::Get(v8::Localv8::Context, v8::Localv8::Value)’
3717 | V8_WARN_UNUSED_RESULT MaybeLocal Get(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3717:43: note: candidate expects 2 arguments, 1 provided
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3720:43: note: candidate: ‘v8::MaybeLocalv8::Value v8::Object::Get(v8::Localv8::Context, uint32_t)’
3720 | V8_WARN_UNUSED_RESULT MaybeLocal Get(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3720:43: note: candidate expects 2 arguments, 1 provided
In file included from ../src/node-boilerplate/inc/scrypt_async.h:28,
from ../src/node-boilerplate/inc/scrypt_params_async.h:28,
from ../src/node-boilerplate/scrypt_params_async.cc:4:
../src/node-boilerplate/inc/scrypt_common.h:40:48: error: no matching function for call to ‘v8::Object::Get(v8::Localv8::String)’
40 | r(obj->Get(Nan::New("r").ToLocalChecked())->Uint32Value()),
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3717:43: note: candidate: ‘v8::MaybeLocalv8::Value v8::Object::Get(v8::Localv8::Context, v8::Localv8::Value)’
3717 | V8_WARN_UNUSED_RESULT MaybeLocal Get(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3717:43: note: candidate expects 2 arguments, 1 provided
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3720:43: note: candidate: ‘v8::MaybeLocalv8::Value v8::Object::Get(v8::Localv8::Context, uint32_t)’
3720 | V8_WARN_UNUSED_RESULT MaybeLocal Get(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3720:43: note: candidate expects 2 arguments, 1 provided
In file included from ../src/node-boilerplate/inc/scrypt_async.h:28,
from ../src/node-boilerplate/inc/scrypt_params_async.h:28,
from ../src/node-boilerplate/scrypt_params_async.cc:4:
../src/node-boilerplate/inc/scrypt_common.h:41:48: error: no matching function for call to ‘v8::Object::Get(v8::Localv8::String)’
41 | p(obj->Get(Nan::New("p").ToLocalChecked())->Uint32Value()) {}
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3717:43: note: candidate: ‘v8::MaybeLocalv8::Value v8::Object::Get(v8::Localv8::Context, v8::Localv8::Value)’
3717 | V8_WARN_UNUSED_RESULT MaybeLocal Get(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3717:43: note: candidate expects 2 arguments, 1 provided
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3720:43: note: candidate: ‘v8::MaybeLocalv8::Value v8::Object::Get(v8::Localv8::Context, uint32_t)’
3720 | V8_WARN_UNUSED_RESULT MaybeLocal Get(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3720:43: note: candidate expects 2 arguments, 1 provided
In file included from ../src/node-boilerplate/inc/scrypt_params_async.h:28,
from ../src/node-boilerplate/scrypt_params_async.cc:4:
../src/node-boilerplate/inc/scrypt_async.h: In member function ‘virtual void ScryptAsyncWorker::HandleErrorCallback()’:
../src/node-boilerplate/inc/scrypt_async.h:53:29: warning: ‘v8::Localv8::Value Nan::Callback::Call(int, v8::Localv8::Value) const’ is deprecated [-Wdeprecated-declarations]
53 | callback->Call(1, argv);
| ^
In file included from ../src/node-boilerplate/scrypt_params_async.cc:1:
../../nan/nan.h:1674:3: note: declared here
1674 | Call(int argc, v8::Localv8::Value argv[]) const {
| ^~~~
In file included from ../src/node-boilerplate/scrypt_params_async.cc:4:
../src/node-boilerplate/inc/scrypt_params_async.h: In constructor ‘ScryptParamsAsyncWorker::ScryptParamsAsyncWorker(Nan::NAN_METHOD_ARGS_TYPE)’:
../src/node-boilerplate/inc/scrypt_params_async.h:35:36: error: no matching function for call to ‘v8::Value::NumberValue()’
35 | maxtime(info[0]->NumberValue()),
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:2861:39: note: candidate: ‘v8::Maybe v8::Value::NumberValue(v8::Localv8::Context) const’
2861 | V8_WARN_UNUSED_RESULT Maybe NumberValue(Local context) const;
| ^~~~~~~~~~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:2861:39: note: candidate expects 1 argument, 0 provided
In file included from ../src/node-boilerplate/scrypt_params_async.cc:4:
../src/node-boilerplate/inc/scrypt_params_async.h:36:39: error: no matching function for call to ‘v8::Value::NumberValue()’
36 | maxmemfrac(info[1]->NumberValue()),
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:2861:39: note: candidate: ‘v8::Maybe v8::Value::NumberValue(v8::Localv8::Context) const’
2861 | V8_WARN_UNUSED_RESULT Maybe NumberValue(Local context) const;
| ^~~~~~~~~~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:2861:39: note: candidate expects 1 argument, 0 provided
In file included from ../src/node-boilerplate/scrypt_params_async.cc:4:
../src/node-boilerplate/inc/scrypt_params_async.h:37:36: error: no matching function for call to ‘v8::Value::IntegerValue()’
37 | maxmem(info[2]->IntegerValue()),
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:2863:40: note: candidate: ‘v8::Maybe v8::Value::IntegerValue(v8::Localv8::Context) const’
2863 | V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
| ^~~~~~~~~~~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:2863:40: note: candidate expects 1 argument, 0 provided
In file included from ../src/node-boilerplate/scrypt_params_async.cc:4:
../src/node-boilerplate/inc/scrypt_params_async.h:38:39: error: no matching function for call to ‘v8::Value::IntegerValue()’
38 | osfreemem(info[3]->IntegerValue())
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:2863:40: note: candidate: ‘v8::Maybe v8::Value::IntegerValue(v8::Localv8::Context) const’
2863 | V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
| ^~~~~~~~~~~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:2863:40: note: candidate expects 1 argument, 0 provided
../src/node-boilerplate/scrypt_params_async.cc: In member function ‘virtual void ScryptParamsAsyncWorker::HandleOKCallback()’:
../src/node-boilerplate/scrypt_params_async.cc:23:67: error: no matching function for call to ‘v8::Object::Set(v8::Localv8::String, Nan:imp:IntegerFactoryv8::Integer::return_t)’
23 | obj->Set(Nan::New("N").ToLocalChecked(), Nan::New(logN));
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3670:37: note: candidate: ‘v8::Maybe v8::Object::Set(v8::Localv8::Context, v8::Localv8::Value, v8::Localv8::Value)’
3670 | V8_WARN_UNUSED_RESULT Maybe Set(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3670:37: note: candidate expects 3 arguments, 2 provided
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3673:37: note: candidate: ‘v8::Maybe v8::Object::Set(v8::Localv8::Context, uint32_t, v8::Localv8::Value)’
3673 | V8_WARN_UNUSED_RESULT Maybe Set(Local context, uint32_t index,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3673:37: note: candidate expects 3 arguments, 2 provided
../src/node-boilerplate/scrypt_params_async.cc:24:64: error: no matching function for call to ‘v8::Object::Set(v8::Localv8::String, Nan:imp:IntegerFactoryv8::Integer::return_t)’
24 | obj->Set(Nan::New("r").ToLocalChecked(), Nan::New(r));
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3670:37: note: candidate: ‘v8::Maybe v8::Object::Set(v8::Localv8::Context, v8::Localv8::Value, v8::Localv8::Value)’
3670 | V8_WARN_UNUSED_RESULT Maybe Set(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3670:37: note: candidate expects 3 arguments, 2 provided
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3673:37: note: candidate: ‘v8::Maybe v8::Object::Set(v8::Localv8::Context, uint32_t, v8::Localv8::Value)’
3673 | V8_WARN_UNUSED_RESULT Maybe Set(Local context, uint32_t index,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3673:37: note: candidate expects 3 arguments, 2 provided
../src/node-boilerplate/scrypt_params_async.cc:25:64: error: no matching function for call to ‘v8::Object::Set(v8::Localv8::String, Nan:imp:IntegerFactoryv8::Integer::return_t)’
25 | obj->Set(Nan::New("p").ToLocalChecked(), Nan::New(p));
| ^
In file included from /home/josh/.cache/node-gyp/14.16.1/include/node/node.h:67,
from ../../nan/nan.h:53,
from ../src/node-boilerplate/scrypt_params_async.cc:1:
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3670:37: note: candidate: ‘v8::Maybe v8::Object::Set(v8::Localv8::Context, v8::Localv8::Value, v8::Localv8::Value)’
3670 | V8_WARN_UNUSED_RESULT Maybe Set(Local context,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3670:37: note: candidate expects 3 arguments, 2 provided
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3673:37: note: candidate: ‘v8::Maybe v8::Object::Set(v8::Localv8::Context, uint32_t, v8::Localv8::Value)’
3673 | V8_WARN_UNUSED_RESULT Maybe Set(Local context, uint32_t index,
| ^~~
/home/josh/.cache/node-gyp/14.16.1/include/node/v8.h:3673:37: note: candidate expects 3 arguments, 2 provided
../src/node-boilerplate/scrypt_params_async.cc:32:25: warning: ‘v8::Localv8::Value Nan::Callback::Call(int, v8::Localv8::Value) const’ is deprecated [-Wdeprecated-declarations]
32 | callback->Call(2, argv);
| ^
In file included from ../src/node-boilerplate/scrypt_params_async.cc:1:
../../nan/nan.h:1674:3: note: declared here
1674 | Call(int argc, v8::Localv8::Value argv[]) const {
| ^~~~
make: *** [scrypt.target.mk:131: Release/obj.target/scrypt/src/node-boilerplate/scrypt_params_async.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: make failed with exit code: 2
gyp ERR! stack at ChildProcess.onExit (/home/josh/.nvm/versions/node/v14.16.1/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack at ChildProcess.emit (events.js:315:20)
gyp ERR! stack at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)
gyp ERR! System Linux 5.4.0-26-generic
gyp ERR! command "/home/josh/.nvm/versions/node/v14.16.1/bin/node" "/home/josh/.nvm/versions/node/v14.16.1/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/josh/.nvm/versions/node/v14.16.1/lib/node_modules/loom-truffle-provider/node_modules/scrypt
gyp ERR! node -v v14.16.1
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! scrypt@6.0.3 install: node-gyp rebuild
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the scrypt@6.0.3 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR! /home/josh/.npm/_logs/2021-04-23T18_44_51_702Z-debug.log

**Comment by joshuaebowling** [comment]
for whatever reason, I solved the same issue by targeting loom-js@1.75.2

running: 
ubuntu 20.04
node 14.16.1

**Comment by shashi278** [comment]
@joshuaebowling I see you had the same issue related to scrypt@6.0.3 and that's bcz scrypt@6.0.3 doesn't support higher node versions so npm installing `github:barrysteyn/node-scrypt#fb60a8d3c158fe115a624b5ffa7480f3a24b03fb` should have solved it as well

**Comment by joshuaebowling** [comment]
@shashi278 
I'm sure you are correct, this is probably what I'll do in the near term if this doesn't resolve itself. 
For now, I did what I wrote above then forked loom-truffle-provider and changed the version of loom-js to 1.75.2 and all is well now.
Thanks!

**Comment by enlight** [comment]
Any `scrypt` related issues should hopefully be resolved in `loom-js@1.79.0` and `loom-truffle-provider@0.15.0`.
